File: zig/linked_list.zig
pub fn LinkedList(comptime T: type) type {
    return struct {
        pub const Node = struct {
            prev: ?*Node = null,
            next: ?*Node = null,
            data: T,
        };

        first: ?*Node = null,
        last: ?*Node = null,
        len: usize = 0,

        const Self = @This();

        pub fn push(self: *Self, node: *Node) void {
            node.prev = self.last;

            if (self.last) |tail| {
                tail.next = node;
            } else {
                self.first = node;
            }
            self.last = node;
            self.len += 1;
        }

        pub fn pop(self: *Self) ?*Node {
            if (self.len == 0) {
                return null;
            }
            self.len -= 1;
            const node = self.last;
            self.last = node.?.prev;
            if (self.last) |tail| {
                tail.next = null;
            } else {
                self.first = null;
            }
            return node;
        }

        pub fn shift(self: *Self) ?*Node {
            if (self.len == 0) {
                return null;
            }
            self.len -= 1;
            const node = self.first;
            self.first = node.?.next;
            if (self.first) |head| {
                head.prev = null;
            } else {
                self.last = null;
            }
            return node;
        }

        pub fn unshift(self: *Self, node: *Node) void {
            node.next = self.first;
            if (self.first) |head| {
                head.prev = node;
            } else {
                self.last = node;
            }
            self.first = node;
            self.len += 1;
        }

        pub fn delete(self: *Self, node: *Node) void {
            var curr: ?*Node = self.first;
            while (curr) |n| : (curr = n.next) {
                if (n == node) {
                    break;
                }
            }
            if (curr) |n| {
                self.len -= 1;
                const prev = n.prev;
                const next = n.next;
                if (prev) |p| {
                    p.next = next;
                } else {
                    self.first = next;
                }
                if (next) |p| {
                    p.prev = prev;
                } else {
                    self.last = prev;
                }
            }
        }
    };
}

File: go/linked_list.go
package linkedlist

import "errors"

type Node struct {
	Value any
	prev  *Node
	next  *Node
}

type List struct {
	head *Node
	tail *Node
}

func NewList(elements ...interface{}) *List {
	l := &List{}
	for _, elem := range elements {
		l.Push(elem)
	}
	return l
}

func (n *Node) Next() *Node {
	return n.next
}

func (n *Node) Prev() *Node {
	return n.prev
}

func (l *List) Unshift(v interface{}) {
	node := &Node{}
	node.Value = v
	if l.head == nil {
		l.tail = node
	} else {
		l.head.prev = node
		node.next = l.head
	}

	l.head = node
}

func (l *List) Push(v interface{}) {
	node := &Node{}
	node.Value = v
	if l.head == nil {
		l.head = node
	} else {
		l.tail.next = node
		node.prev = l.tail
	}

	l.tail = node
}

func (l *List) Shift() (interface{}, error) {
	if l.head == nil {
		return nil, errors.New("empty")
	}

	node := l.head
	l.head = node.next
	if l.head == nil {
		l.tail = nil
	} else {
		l.head.prev = nil
	}
	return node.Value, nil
}

func (l *List) Pop() (interface{}, error) {
	if l.head == nil {
		return nil, errors.New("empty")
	}

	node := l.tail
	l.tail = node.prev
	if l.tail == nil {
		l.head = nil
	} else {
		l.tail.next = nil
	}
	return node.Value, nil
}

func (l *List) Reverse() {
	if l.head == nil {
		return
	}

	for curr := l.head; curr != nil; curr = curr.prev {
		curr.next, curr.prev = curr.prev, curr.next
	}

	l.head, l.tail = l.tail, l.head
}

func (l *List) First() *Node {
	return l.head
}

func (l *List) Last() *Node {
	return l.tail
}
